### PR TITLE
Filter early on buffer id and pass vector of entries in normalize_values

### PR DESCRIPTION
**Pull request by toregge**
and foreach_value ShardedHashMap member functions to limit number of callbacks.

@geirst : please review

### CODE REVIEW

File: searchlib/src/vespa/searchlib/attribute/postingstore.cpp [hunk 1]
```diff
@@ -647,86 +651,92 @@ PostingStore<DataT>::update_stat()
...
 void
 PostingStore<DataT>::compact_worst_btree_nodes()
 {
     auto to_hold = this->start_compact_worst_btree_nodes();
-    _dictionary.normalize_posting_lists([this](EntryRef posting_idx) -> EntryRef
-                                        {
-                                            move_btree_nodes(posting_idx);
-                                            return posting_idx;
-                                        });
+    std::vector<bool> filter(RefType::numBuffers());
```
Comment (geirst) [inline searchlib/src/vespa/searchlib/attribute/postingstore.cpp:730]: Please create a helper function to create the filter.

**Review by geirst — APPROVED** [review]
👍 But I think we should simplify how to represent the filter by creating a separate class for that. E.g. EntryRefFilter, that contains the buffer id filter vector and entry_ref_offset_bits. It can have _bool contains(EntryRef ref) const_ that we use inside the normalize_xxx() functions, and _void add_buffer(uint32_t buffer_id)_ and _void add_buffers(const std::vector<uint32_t>& ids)_ that are used to populate the filter. This will also simplify the signature of the _normalize_xxx()_ functions.

File: vespalib/src/vespa/vespalib/datastore/fixed_size_hash_map.h [hunk 2]
```diff
@@ -160,6 +160,8 @@ class FixedSizeHashMap {
     void foreach_key(const std::function<void(EntryRef)>& callback) const;
     void move_keys(ICompactable& compactable, const std::vector<bool>& compacting_buffers, uint32_t entry_ref_offset_bits);
     bool normalize_values(const std::function<EntryRef(EntryRef)>& normalize);
+    bool normalize_values(const std::function<void(std::vector<EntryRef>&)>& normalize, const std::vector<bool>& filter, uint32_t entry_ref_offset_bits);
```
[the remark this answers is not in the document]
Comment (toregge) [reply vespalib/src/vespa/vespalib/datastore/fixed_size_hash_map.h]: Done.

File: searchlib/src/vespa/searchlib/attribute/i_enum_store_dictionary.h [hunk 3]
```diff
@@ -53,6 +53,8 @@ class IEnumStoreDictionary : public vespalib::datastore::IUniqueStoreDictionary
     virtual void clear_all_posting_lists(std::function<void(EntryRef)> clearer) = 0;
     virtual void update_posting_list(Index idx, const EntryComparator& cmp, std::function<EntryRef(EntryRef)> updater) = 0;
     virtual bool normalize_posting_lists(std::function<EntryRef(EntryRef)> normalize) = 0;
+    virtual bool normalize_posting_lists(std::function<void(std::vector<EntryRef>&)> normalize, const std::vector<bool>& filter, uint32_t entry_ref_offset_bits) = 0;
```
[the remark this answers is not in the document]
Comment (toregge) [reply searchlib/src/vespa/searchlib/attribute/i_enum_store_dictionary.h]: Done.

File: searchlib/src/tests/attribute/enumstore/enumstore_test.cpp [hunk 4]
```diff
@@ -621,43 +639,146 @@ EnumStoreDictionaryTest<EnumStoreTypeAndDictionaryType>::insert_value(size_t val
...
-    EXPECT_EQ(EntryRef(), find_result.second);
+    EXPECT_EQ(exp_adjusted_refs, get_sample_values(large_population));
+    clear_sample_values(large_population);
 }
 
+template <typename EnumStoreTypeAndDictionaryType>
+void
+EnumStoreDictionaryTest<EnumStoreTypeAndDictionaryType>::test_foreach_posting_list(bool one_filter)
+{
+    std::vector<bool> filter;
```
[the remark this answers is not in the document]
Comment (toregge) [reply searchlib/src/tests/attribute/enumstore/enumstore_test.cpp]: Fixed.

File: searchlib/src/vespa/searchlib/attribute/postingstore.cpp (hunk 1 above)
Comment (toregge) [reply searchlib/src/vespa/searchlib/attribute/postingstore.cpp]: Fixed.

File: searchlib/src/vespa/searchlib/attribute/postingstore.cpp [hunk 5]
```diff
@@ -735,8 +745,25 @@ void
 PostingStore<DataT>::compact_worst_buffers()
 {
     auto to_hold = this->start_compact_worst_buffers();
-    _dictionary.normalize_posting_lists([this](EntryRef posting_idx) -> EntryRef
-                                        { return move(posting_idx); });
+    bool compact_btree_roots = false;
+    std::vector<bool> filter(RefType::numBuffers());
```
[the remark this answers is not in the document]
Comment (toregge) [reply searchlib/src/vespa/searchlib/attribute/postingstore.cpp]: Fixed.

File: searchlib/src/tests/attribute/enumstore/enumstore_test.cpp [hunk 6]
```diff
@@ -621,43 +639,146 @@ EnumStoreDictionaryTest<EnumStoreTypeAndDictionaryType>::insert_value(size_t val
...
+        exp_refs.emplace_back(make_fake_pidx(i));
+        if (!use_filter || !one_filter || select_buffer(i) == 3) {
+            exp_adjusted_refs.emplace_back(make_fake_adjusted_pidx(i));
+        } else {
+            exp_adjusted_refs.emplace_back(make_fake_pidx(i));
+        }
+    }
+    EXPECT_EQ(exp_refs, get_sample_values(large_population));
     if (use_filter) {
+        std::vector<bool> filter;
```
[the remark this answers is not in the document]
Comment (toregge) [reply searchlib/src/tests/attribute/enumstore/enumstore_test.cpp]: Fixed.

**Review by geirst — APPROVED** [review]
👍

**Pull request merged**